### ISSUE TITLE
mdns is squating on host hostname

### ISSUE THREAD

**Issue by daviddias** [libp2p/go-libp2p#176]
I've been debugging why jsipfs and go-ipfs stopped finding each other since we got it to work (end of 2015).

It seems that couple of things have changed in go-ipfs, starting with the `serviceTag` and the really critical change, `go-ipfs` now squats the host hostname to announce its records through mdns.

See that https://github.com/libp2p/go-libp2p/blob/master/p2p/discovery/mdns.go#L85 passes `""` and then https://github.com/whyrusleeping/mdns/blob/master/zone.go#L83-L91

This is the reason why I keep seeing my hostname change when I have an `ipfs daemon` running:
![image](https://cloud.githubusercontent.com/assets/1211152/22023635/de2c83d4-dcbe-11e6-8bba-a7ea285934e5.png)

What happens is that (by definition of the mdns protocol), once there is another announcer using the same hostname, one of them has to change.

Unless I'm missing something, I don't see a need to take over the host hostname. 

PS: This 'fault' in the protocol is what enabled Watson to hack printers :) https://www.youtube.com/watch?v=BMWP20ZicsY

**Comment by diasdavid** [comment]
@whyrusleeping mind taking a look? I've reached out to @Kubuxu but he was unsure.

**Comment by whyrusleeping** [comment]
I guess we could try passing the peer ID as the 'hostname' there, that should work right?

**Comment by diasdavid** [comment]
It needs to be a common thing across nodes, otherwise, they won't know whom to query. What about just `ipfs.local`?

**Comment by wigy-opensource-developer** [comment]
I could imagine people running multiple IPFS nodes in a broadcast domain. Using `ipfs.local` as the hostname would hurt that use-case.

**Comment by diasdavid** [comment]
Back in the 'old days', when ipfs mdns interop worked, this was what go-ipfs used to send as an answer to `discovery.ipfs.io.local`

```
   [ { name: 'discovery.ipfs.io.local',
       type: 'PTR',
       class: 1,
       ttl: 120,
       data: 'QmbBHw1Xx9pUpAbrVZUKTPL5Rsph5Q9GQhRvcWVBPFgGtC.discovery.ipfs.io.local' },

     { name: 'QmbBHw1Xx9pUpAbrVZUKTPL5Rsph5Q9GQhRvcWVBPFgGtC.discovery.ipfs.io.local',
       type: 'SRV',
       class: 1,
       ttl: 120,
       data: { priority: 10, weight: 1, port: 4001, target: 'lorien.local' } },

     { name: 'lorien.local',
       type: 'A',
       class: 1,
       ttl: 120,
       data: '127.0.0.1' },

     { name: 'lorien.local',
       type: 'A',
       class: 1,
       ttl: 120,
       data: '127.94.0.1' },

     { name: 'lorien.local',
       type: 'A',
       class: 1,
       ttl: 120,
       data: '172.16.38.224' },

     { name: 'QmbBHw1Xx9pUpAbrVZUKTPL5Rsph5Q9GQhRvcWVBPFgGtC.discovery.ipfs.io.local',
       type: 'TXT',
       class: 1,
       ttl: 120,
       data: 'QmbBHw1Xx9pUpAbrVZUKTPL5Rsph5Q9GQhRvcWVBPFgGtC' } ],
```

Listening for queries on ipfs.local would be enough, then when replying, we can do <peer-id>-ipfs.local

**Comment by whyrusleeping** [comment]
wait, if it needs to be common across nodes and we're setting it to `os.Hostname` currently, how does it work at all?

**Comment by diasdavid** [comment]
It squats it

> PS: This 'fault' in the protocol is what enabled Watson to hack printers :) https://www.youtube.com/watch?v=BMWP20ZicsY

**Comment by lgierth** [comment]
@diasdavid how can I reproduce this with js-ipfs? (I'm pretty sure I won't be able to reproduce it on Linux, can't change the hostname as unprivileged user.)

**Comment by dignifiedquire** [comment]
any updates on this?

**Comment by lgierth** [comment]
I had a hard time reproducing it and moved to other things

**Comment by diasdavid** [comment]
@lgierth you should be looking into reproducing it with go-ipfs, that is where I get it to happen, although I'm trying to reproduce it today and it is not happening (or at least not as fast), will continue investigating.

**Comment by diasdavid** [comment]
@richardschneider you have done a lot of work towards fixing mdns in both js and go. Could you share an update here on the current status? Thank you in advance.

**Comment by richardschneider** [comment]
https://github.com/libp2p/js-libp2p-mdns now uses `ipfs.local` as the service tag for the query.  The fully qualified server name is `Qm...ipfs.local`.

There are no resources to work on the GO implementation, https://github.com/libp2p/js-libp2p-mdns.

I'll update the JS readme to give an example query and response.

**Comment by lidel** [comment]
In case someone find this issue, the latest "mdns" discovery spec is at [libp2p/specs/discovery/mdns.md](https://github.com/libp2p/specs/blob/master/discovery/mdns.md)
As I type this both old and new specs are implemented in js-ipfs 0.50, but only old one is supported in go-ipfs 0.7.
